tree-sitter-stack-graphs/src/lib.rs

```diff
@@ -550,52 +633,51 @@ impl From<tree_sitter_graph::ExecutionError> for LoadError {
     }
 }
 
-struct StackGraphLoader<'a> {
-    stack_graph: &'a mut StackGraph,
-    file: Handle<File>,
-    graph: &'a Graph<'a>,
-    source: &'a str,
-    span_calculator: SpanCalculator<'a>,
-}
+impl<'a> Builder<'a> {
+    fn load(mut self, cancellation_flag: &dyn CancellationFlag) -> Result<(), LoadError> {
+        let cancellation_flag: &dyn stack_graphs::CancellationFlag = &cancellation_flag;
 
-impl<'a> StackGraphLoader<'a> {
-    fn new(
-        stack_graph: &'a mut StackGraph,
-        file: Handle<File>,
-        graph: &'a Graph<'a>,
-        source: &'a str,
-    ) -> Self {
-        let span_calculator = SpanCalculator::new(source);
-        StackGraphLoader {
-            stack_graph,
-            file,
-            graph,
-            source,
-            span_calculator,
+        // By default graph ids are used for stack graph local_ids. A remapping is computed
+        // for local_ids that already exist in the graph---all other graph ids are mapped to
+        // the same local_id.
+        let mut next_local_id = (self.graph.node_count() - self.injected_node_count) as u32;
```

Comment:
It works as follows. The `Graph` contains `i` injected nodes `0..i-1`, and `n` nodes `i..(i+n)-1` that are created during execution. The latter are mapped to local ids `0..n-1`. The `x` local ids that already exist in the graph are remapped to `n..(n+x)-1`.

The idea is that only local ids that already exist are remapped and thus get an entry in the map. I expect that the set of preexisting nodes is usually small, so only a few local ids are remapped. There can be gaps in the local ids if the TSG has created nodes without attributes that make them result in stack graph nodes.

I briefly considered other schemes like your suggestion to start the search for available local ids from 0. This can easily result in a situation where a preexisting node with local id `0` results in all local ids being shifted by 1 (0 -> 1, then 1 must go to 2, etc). I couldn't quickly think of a method that would prevent that.

Reply:
Oh I didn't realize the shifting part was happening.  Is that so that the [`GraphNodeID`](https://github.com/tree-sitter/tree-sitter-graph/blob/1c99c2c6b7edf86eac74941f042a0c4f3f2411a2/src/graph.rs#L44) from `tree-sitter-graph` lines up with the local node ID in the stack graph that we create?  Technically that's not needed as a requirement, except for that it would make it easier to compare e.g. tree-sitter-graph's JSON output (which will show `GraphNodeID`s) with the stack graphs produced here (which will show stack graph local node IDs).  So you could simplify this by not requiring that consistency.

I'm also okay leaving it how it is, but I think this deserves a big explanatory comment somewhere explaining (a) the requirement and (b) how this logic satisfies that requirement.